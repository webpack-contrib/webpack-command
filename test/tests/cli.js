const execa = require('execa');
const strip = require('strip-ansi');

const { test } = require('../util');

test('Bad Config', module, () => {
  it('should run', () => {
    const cliPath = resolve(__dirname, '../../lib/cli');
    const cwd = resolve(__dirname, '../fixtures/bad-config');

    try {
      execa.sync('node', [cliPath], { cwd });
    } catch (e) {
      expect(e.message).toMatch(
        `options['batman']  is an invalid additional property`
      );
    }
  });
});

test('Zero Config', module, () => {
  it('should run', () => {
    const cliPath = resolve(__dirname, '../../lib/cli');
    const cwd = resolve(__dirname, '../fixtures/zero-config');
    const result = execa.sync('node', [cliPath], { cwd });

    expect(
      strip(result.stdout)
        .replace(/Δt \d+ms/g, '<duration>')
        .replace(/v\d\.\d\.\d/, '<version>')
    ).toMatchSnapshot();
  }).timeout(4000);
});

test('Commands', module, () => {
  it('should show cli help', () => {
    const cliPath = resolve(__dirname, '../../lib/cli.js');
    const result = execa.sync('node', [cliPath, '--help']);

    expect(
      strip(result.stdout).replace(/Δt \d+ms/g, '<duration>')
    ).toMatchSnapshot();
  });

  it('should show teach command help', () => {
    const cliPath = resolve(__dirname, '../../lib/cli.js');
    const result = execa.sync(cliPath, ['help', 'teach']);

    expect(result.stdout.replace(/v\d\.\d\.\d/, '<version>')).toMatchSnapshot();
  }).timeout(4000);

  it('should throw on command error', () => {
    const cliPath = resolve(__dirname, '../../lib/cli.js');
    const stub = () => execa.sync(cliPath, ['help', 'bad-command']);

    // TODO: move to error codes
    expect(stub).toThrow(/'bad-command' has not been installed/);
  }).timeout(4000);

  it('should throw on missing command', () => {
    const cliPath = resolve(__dirname, '../../lib/cli.js');
    const stub = () =>
      execa.sync(cliPath, ['bad-command'], {
        env: {
          CLI_TEST: 'true',
        },
      });

    // TODO: move to error codes
    expect(stub).toThrow(/bad-command/);
  }).timeout(4000);

  it('should add directories to entries', () => {
    const cliPath = resolve(__dirname, '../../lib/cli.js');
    const srcPath = resolve(__dirname, '../../test/fixtures/flags/config/src');
    const result = execa.sync(cliPath, [srcPath]);

    expect(
      strip(result.stdout).replace(/Δt \d+ms/g, '<duration>')
    ).toMatchSnapshot();
  }).timeout(4000);

  it('should accept custom reporters relative to the current working directory', () => {
    const cliPath = resolve(__dirname, '../../lib/cli.js');
    const srcPath = resolve(__dirname, '../../test/fixtures/flags/config/src');
    const result = execa.sync(
      cliPath,
      ['--reporter', '../../lib/reporters/BasicReporter', srcPath],
      {
        cwd: __dirname,
      }
    );

    expect(
      strip(result.stdout)
        .replace(/Version: webpack \d\.\d\.\d/, '<version>')
        .replace(/Time: \d+ms/g, '<duration>')
        .replace(/Built at: .+(?=\n)/g, '<datetime>')
    ).toMatchSnapshot();
  }).timeout(4000);
});
