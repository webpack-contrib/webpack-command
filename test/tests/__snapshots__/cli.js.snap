// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands > should add directories to entries #0 1`] = `
"‚Ñπ ÔΩ¢webpackÔΩ£: Starting Build
‚Ñπ ÔΩ¢webpackÔΩ£: Build Finished

webpack v4.6.0

953ce188d35af8c002f0
  size   name  module                                     status
  43 B   0     ./test/fixtures/flags/config/src/index.js  built

  size   name  asset                                      status
  573 B  main  main.js                                    emitted

  <duration> 


configuration
  0:0  warning  The 'mode' option has not been set, webpack will fallback to
                'production' for this value. Set 'mode' option to 'development' or
                'production' to enable defaults for each environment.

‚ö†  1 problem (0 errors, 1 warning)"
`;

exports[`Commands > should show cli help #0 1`] = `
"
  üêï A lightweight, modular, and opinionated webpack CLI

  Usage
    $ webpack [<config>, ...options]
    $ webpack <entry-file> [...<entry-file>] <output-file>

  Options
    --context                     The root directory for resolving entry point and stats
    --debug                       Switch loaders to debug mode
    --devtool                     Enable devtool for better debugging experience.
                                  e.g. --devtool eval-cheap-module-source-map
    --entry                       The entry point
    --help                        Show usage information and the options listed here
    --log-level                   Limit all process console messages to a specific level and above
                                  Levels: trace, debug, info, warn, error, silent
    --log-time                    Instruct the logger and dependencies to display a timestamp
    --progress                    Instructs webpack to track and display build progress
                                  This is often used with --profile
    --reporter                    Specifies the reporter to use for generating console output for a build
    --require                     Preload one or more modules before loading the webpack configuration
                                  Typically used for language-specific require hooks
    --run-dev                     An alias for --debug --devtool eval-cheap-module-source-map
                                  --mode development --output-pathinfo
    --run-prod                    An alias for --optimize-minimize --mode production, and defines
                                  process.env.NODE_ENV=\\"production\\"
    --version                     Display the webpack-command version
    --watch                       Watch the filesystem for changes

  Advanced
    --bail                        Abort the compilation on first error
    --cache                       Enable in memory caching
    --define                      Used to redefine or replace a variable or key in the bundle. Complex
                                  keys should be set in config. {dim e.g. --define.batman robin}
    --hot                         Enables Hot Module Replacement
    --plugin                      Load this plugin
    --prefetch                    Prefetch this request
                                  e.g. --prefetch ./file.js
    --profile                     Profile the compilation and include information in stats
    --provide                     Provide these modules as free vars in all modules
                                  e.g. --provide.jQuery jquery
    --records-input-path          Path to the records file (reading)
    --records-output-path         Path to the records file (writing)
    --records-path                Path to the records file
    --target                      The targeted execution environment
    --watch-aggregate-timeout     Timeout for gathering changes while watching
    --watch-poll                  The polling interval for watching (also enable polling)
    --watch-stdin                 Exit the process when stdin is closed

  Configuration File
    --config                      Path to the config file
    --config-name                 Name of the config to use
    --config-register             Deprecated. Please use --require.

    --mode                        Specifies the build mode to use; development or production

  Modules
    --module-bind                 Bind an extension to a loader
    --module-bind-post            Bind an extension to a postLoader
    --module-bind-pre             Bind an extension to a preLoader

  Optimization
    --optimize-max-chunks         Try to keep the chunk count below a limit
    --optimize-min-chunk-size     Try to keep the chunk size above a limit
    --optimize-minimize           Minimize javascript and switches loaders to minimizing

  Output
    --output                      The output path and file for compilation assets
    --output-chunk-filename       The output filename for additional chunks
    --output-filename             The output filename of the bundle
    --output-jsonp-function       The name of the JSONP function used for chunk loading
    --output-library              Expose the exports of the entry point as library
    --output-library-target       The type for exposing the exports of the entry point as library
    --output-path                 The output path for compilation assets
    --output-pathinfo             Include a comment with the request for every dependency (require, import, etc.)
    --output-public-path          The public path for the assets
    --output-source-map-filename  The output filename for the SourceMap

  Resolver
    --resolve-alias               Setup a module alias for resolving
                                  e.g. --resolve-alias.jquery jquery.plugin
    --resolve-extensions          Setup extensions that should be used to resolve modules
                                  e.g. .es6,.js
    --resolve-loader-alias        Setup a loader alias for resolving


    For further documentation, visit https://webpack.js.org/api/cli

  Commands
    help
    teach

    Type \`webpack help <command>\` for more information

  Examples
    $ webpack
    $ webpack --help
    $ webpack entry.js
    $ webpack --config ../webpack.config.js
"
`;

exports[`Commands > should show teach command help #0 1`] = `
"
  teach <version>

  Teaches webpack-command that a command has been installed and is available.

  Usage
    $ webpack teach --command <command> --module <module>

  Options
    --command   The name of a command that users will type
    --forget    Instructs the tool to forget a previously added command
    --module    The npm module name of a command

  Examples
    $ webpack teach --command init --module webpack-command-init
    $ webpack teach --command init --forget
"
`;

exports[`Zero Config > should run #0 1`] = `
"‚Ñπ ÔΩ¢webpackÔΩ£: Starting Build
‚Ñπ ÔΩ¢webpackÔΩ£: Build Finished

webpack <version>

953ce188d35af8c002f0
  size   name  module          status
  43 B   0     ./src/index.js  built

  size   name  asset           status
  573 B  main  main.js         emitted

  <duration> 


configuration
  0:0  warning  The 'mode' option has not been set, webpack will fallback to
                'production' for this value. Set 'mode' option to 'development' or
                'production' to enable defaults for each environment.

‚ö†  1 problem (0 errors, 1 warning)"
`;
